clippy_lints/src/dereference.rs

```diff
@@ -239,14 +244,21 @@ impl<'tcx> LateLintPass<'tcx> for Dereferencing<'tcx> {
             return;
         }
 
-        if let Some(local) = expr.res_local_id() {
+        if let Some(local) = path_to_local(expr) {
             self.check_local_usage(cx, expr, local);
         }
 
         // Stop processing sub expressions when a macro call is seen
         if expr.span.from_expansion() {
             if let Some((state, data)) = self.state.take() {
-                report(cx, expr, state, data, cx.typeck_results());
+                report(
+                    cx,
+                    expr,
+                    state,
+                    data,
+                    cx.typeck_results(),
+                    self.in_deref_or_deref_mut_impl,
```

Comment:
As soon as you need to pass a `self` field to the `report` function, it might make sense to make it a method instead and remove this extra argument.